### PR TITLE
[IA-2679][IA-2814] UI support for creating GCE VMs with GPUs

### PR DESCRIPTION
**Pull request by kyuksel**
**Note:** This PR needs to be merged after other GPU support dependencies are in place.

## [IA-2814](https://broadworkbench.atlassian.net/browse/IA-2814)
image

Small amount of non-functional changes to fix wording on the window above around Galaxy pause, resume and auto-pause functionality. Please see the ticket for details. In hindsight, I should've PRed these changes separately but things have gone differently than I'd anticipated.

## [IA-2679](https://broadworkbench.atlassian.net/browse/IA-2679)

image

### Requirements
The requirements described in the ticket map to quite a few lower-level scenarios to handle differently for GPUs. I listed them [here](https://broadworkbench.atlassian.net/wiki/spaces/IA/pages/2039906312/2021-06-24+GPU+UI+Requirements). These should also serve as helpful reference while manually testing the PR.

### Known Issue
Google "[projects have a global GPU quota](https://cloud.google.com/compute/docs/gpus#restrictions), which limits the total number of GPUs you can create in any supported zone." Therefore GPU creation requests can fail several times in a row. We are hoping to improve UX experience around this but this is beyond the scope of this PR.

### Remaining TODOs
✅   Don't display `Enable GPUs` when there is a GCE compute without GPUs
✅   Fix spacing such that Spark cluster disk size box fits into the section 
✅   Update `margin` and `padding` across `NewRuntimeModal` as per UX mock
🔲   Update the link to point to the GPU documentation once the User Ed team is done with it

### Beyond Scope & Future Work
- Add MixPanel analytics for GPU usage pattern to inform selections and resource quota issue severity
- Improving UX around failures due to the aforementioned quota issue
- Enable `GPU machine family (A2 series)` which is optimized for GPU use cases
- Improve UI messaging so people can find out more easily all available GPU types and numbers, even when some of them may not be allowed due to number of CPUs and/or memory restrictions
- Creating or patching non-GCE compute (i.e. Dataproc and K8s) with GPUs
- Attaching GPUs to an existing GCE compute
- Patching a GCE compute created with GPUs

### CODE REVIEW

File: src/components/NewGalaxyModal.js
Comment (kyuksel) [inline src/components/NewGalaxyModal.js:168]: There is no functional change in this file.

File: src/components/NewRuntimeModal.js
Comment (kyuksel) [inline src/components/NewRuntimeModal.js:55]: `MachineSelector` used to be used in two different places to select GCE VMs' as well as Dataproc main machines' machine type. Now that GCE VMs can have GPU(s) attached, this component didn't factor well since we want available GPU type and number to dynamically and directly depend on CPU and Memory selections, which doesn't apply to Dataproc backends. Therefore, I renamed this component as (Dataproc) `WorkerSelector` which reflects its only purpose going forward. For the rest, I included the `Select` components flatly in `renderCloudComputeProfileSection()` (formerly `renderRuntimeSection()`).

File: src/components/NewRuntimeModal.js
Comment (kyuksel) [inline src/components/NewRuntimeModal.js:152]: We don't want `Enable GPUs` to be displayed for Dataproc compute and hence `!sparkMode`.

File: src/components/NewRuntimeModal.js
Comment (kyuksel) [inline src/components/NewRuntimeModal.js:153]: We need to behave differently when there's an existing compute with GPUs (more on that later) vs. at compute creation time with no preexisting compute. That's why we have `hasGpu` in addition to `gpuEnabled`.

File: src/components/NewRuntimeModal.js
Comment (kyuksel) [inline src/components/NewRuntimeModal.js:155]: In general, we're defaulting to and ranking more highly in menus the least expensive configurations applicable.

File: src/components/NewRuntimeModal.js
Comment (kyuksel) [inline src/components/NewRuntimeModal.js]: One-`gridStyle`-fit-all wasn't easy to make work since GPU selections have different widths than CPU selections, so I moved `gridTemplateColumns`down to lower-level `div`s for more granular control.

File: src/components/NewRuntimeModal.js
Comment (kyuksel) [inline src/components/NewRuntimeModal.js:783]: These capture the logic that controls dynamic updating of available GPU type and numbers based on CPU number and memory selections. GPU numbers also depend on GPU type. 

I thought about making helper functions for these but they'd be used only here and thought including them here directly may be more reader-friendly in this case.

File: src/data/machines.js
Comment (kyuksel) [inline src/data/machines.js:50]: `name`s are user-facing and `type`s are used internally and by the Leo APIs.

File: src/components/NewRuntimeModal.js
Comment (rtitle) [inline src/components/NewRuntimeModal.js]: Just curious: is there a particular reason this is marked `Beta`? Is there criteria for moving it out of Beta? Just thinking that while there may be evolution of other ML functionality, GPU support specifically is pretty much done.

File: src/components/NewRuntimeModal.js
Comment (dmohs) [inline src/components/NewRuntimeModal.js:191]: Good call moving to single line here.

File: src/components/NewRuntimeModal.js
Comment (dmohs) [inline src/components/NewRuntimeModal.js]: Should this be `(hasGpu && gpuConfig ? { gpuConfig } : {})`?

File: src/components/NewRuntimeModal.js
Comment (kyuksel) [reply src/components/NewRuntimeModal.js]: Good question. I think I'd feel more comfortable with removing `Beta` once we
1. gain more confidence that our revamped container image stack is more proven
2. have a better idea of how much the Google GPU quota issue will affect user experience, and what we can do about it
3. get good feedback from beta testers that it's usable for them

The criteria are admittedly hand-wavy but I think that's okay. 

What do you think?

Comment (rtitle) [reply src/components/NewRuntimeModal.js]: That all sounds good to me 👍 , thanks

**Review by rtitle — APPROVED** [review]
I've been running this branch a lot to test https://github.com/DataBiosphere/terra-docker/pull/209. I've tested creating GPU, non-GPU, and Dataproc runtimes; everything seems to work as expected. It also seems like all requirements from [here](https://broadworkbench.atlassian.net/wiki/spaces/IA/pages/2039906312/2021-06-24+GPU+UI+Requirements) are fulfilled. Nice work!

Note: this should not merge until the above terra-docker PR is merged and released to production.

File: src/components/NewRuntimeModal.js
Comment (dmohs) [inline src/components/NewRuntimeModal.js]:
```suggestion
          div({ style: { ...gridStyle, gridTemplateColumns: '0.25fr 4.5rem 1fr 5.5rem 1fr 5.5rem' } }, [
```
I actually never use backticks in my own code because
```js
`Hello ${name}!`
```
saves barely any room over
```js
'Hello '+name+'!'
```

File: src/components/NewRuntimeModal.js
Comment (dmohs) [inline src/components/NewRuntimeModal.js]:
```suggestion
          gpuEnabled && !sparkMode && div({ style: { ...gridStyle, gridTemplateColumns: '0.75fr 12rem 1fr 5.5rem 1fr 5.5rem', marginTop: '0.75rem' } }, [
```

**Review by dmohs — APPROVED** [review]
Functional approach is really nice across the PR.

File: src/components/NewRuntimeModal.js
Comment (zarsky-broad) [reply src/components/NewRuntimeModal.js]: We do backticks over string concat as a standard in this codebase, especially since we put spaces around operators. That said, this doesn't have any interpolation, so it should indeed be single-quoted.

File: src/libs/runtime-utils.js
Comment (zarsky-broad) [inline src/libs/runtime-utils.js]: Do we need this case if we have the default?

File: src/data/machines.js
Comment (zarsky-broad) [reply src/data/machines.js:50]: Do we need user facing names here if we have `displayNameForGpuType`?

File: src/components/NewRuntimeModal.js
Comment (zarsky-broad) [inline src/components/NewRuntimeModal.js]:
```suggestion
          div({ style: { ...gridStyle, gridTemplateColumns: '0.75fr 4.5rem 1fr 5.5rem 1fr 5.5rem', marginTop: '0.75rem' } }, [
```

File: src/components/NewRuntimeModal.js
Comment (zarsky-broad) [inline src/components/NewRuntimeModal.js]:
```suggestion
          div({ style: gridStyle }, [
```

File: src/components/NewRuntimeModal.js
Comment (zarsky-broad) [reply src/components/NewRuntimeModal.js]: This is an extremely long line, could you break it up? As a guideline, try to avoid lines much over ~120 characters.

File: src/data/machines.js
Comment (kyuksel) [reply src/data/machines.js:50]: We don't need it per se but
- `renderCloudComputeProfileSection()` seems to be called multiple times upon state updates and I wanted to err on the side of optimizing for performance vs. memory assuming a dictionary look-up is a lot quicker than branching through conditionals.
- It also provides a bit more self-documentation I'd say 😊

**Review by kyuksel — COMMENTED** [review]
@zarsky-broad @dmohs @rtitle Thank you very much for the feedback and the testing 🙏🏼 
I've pushed changes to address your comments (or replied to them). A non-white-space diff since your review is [here](https://github.com/DataBiosphere/terra-ui/pull/2554/files/69ede2fb447d47ac01ec5a5904b0f2ecd9ea8515..12dfc0cf7aa65adc0e44bf4bbcb4e68e006f9bfb?w=1) if it helps. The changes also include the following improvements:

- Disabling the `Enable GPUs` checkbox when there is a GCE compute without GPUs, and show a tooltip: This was favoured over not displaying the checkbox at all for UX purposes since the latter can be a bit confusing.
- Fixing a spacing issue where Spark cluster disk size box wasn't fitting into the section.
- Adjusting `margin` and `padding` per the specification received from the UX team below:

gpu2

File: src/data/machines.js
Comment (zarsky-broad) [reply src/data/machines.js:50]: I'd say that's a serious problem that's separate from this, then: a render function should only be called once per update.

Comment (kyuksel) [reply src/data/machines.js:50]: I don't know React well enough to comment on the rendering piece.

If you feel strongly about removing `name`, I can do that.

Comment (zarsky-broad) [reply src/data/machines.js:50]: We can leave it, I just want to deduplicate the calls to the render method.

Comment (kyuksel) [reply src/data/machines.js:50]: I may work on converting `NewRuntimeModal` to be hook-based as part of the jamboree, during which time I may be able to investigate it further.

**Review by rtitle — APPROVED** [review]
Retested basic GPU functionality, it looks good to me. 

Btw to summarize in-person conversations: we're targeting merge/releasing this PR along with Leo next week to release GPUs in Terra.

**Comment by kyuksel** [comment]
[GPU Release Checklist](https://broadworkbench.atlassian.net/wiki/spaces/IA/pages/2071789569/2021-07-01+GPU+Release+Checklist) for reference

**Pull request merged**
